src/cohort/cohort_server.cc

```diff
@@ -1,56 +1,265 @@
 #include "src/cohort/cohort_server.h"
 
+#include <fstream>
+
 #include "absl/status/statusor.h"
+#include "absl/synchronization/mutex.h"
 #include "absl/time/time.h"
 #include "grpcpp/server_context.h"
+#include "src/blockchain/two_phase_commit.h"
 #include "src/proto/cohort.grpc.pb.h"
-#include "src/utils/grpc_utils.h"
+#include "src/utils/status_utils.h"
 
 namespace cohort {
 
 namespace {
 
+using Blockchain = ::blockchain::TwoPhaseCommit;
 using ::grpc::ServerContext;
 
-std::chrono::time_point<std::chrono::system_clock> FromProtoTimestamp(
-    const google::protobuf::Timestamp &timestamp) {
-  return absl::ToChronoTime(absl::FromUnixSeconds(timestamp.seconds()) +
-                            absl::Nanoseconds(timestamp.nanos()));
+Blockchain::VotingDecision WaitForBlockchainDecision(
+    const std::string& transaction_id, absl::Time presumed_abort_time) {
+  std::this_thread::sleep_until(absl::ToChronoTime(presumed_abort_time));
+  Blockchain::VotingDecision decision =
+      Blockchain::GetVotingDecision(transaction_id)
+          .value_or(Blockchain::VotingDecision::PENDING);
+  while (decision == Blockchain::VotingDecision::PENDING) {
+    std::this_thread::sleep_for(
+        absl::ToChronoMicroseconds(absl::Now() - presumed_abort_time));
+    decision = Blockchain::GetVotingDecision(transaction_id)
+                   .value_or(Blockchain::VotingDecision::PENDING);
+  }
+  return decision;
 }
 
 }  // namespace
 
-GetTransactionResultResponse CohortServer::Begin(
-    const PrepareTransactionRequest & /*request*/) {
-  auto response = GetTransactionResultResponse();
-  response.mutable_pending_response();
-  // TODO(benjmarks22): Interact with the database to set the response.
-  // TODO(benjmarks22): Vote in the blockchain based on the database response.
-  return response;
+absl::Mutex& CohortServer::GetLock(const std::string& key) {
```

Comment:
I had impression that we want to use the lock provided by the database. It seems like the database in this case doesn't have to worry about two transaction accessing the same entry?

Reply:
I thought the agreement with @n-triple-a was that it would be simplest if it is handled here since not all databases handle concurrency properly (e.g. the absl::flat_hash_map based "database" I used for testing here).

Reply:
Got it - thanks for the update.

Reply:
+1